Automatic Platform Detection for Development and Distribution

LongLiveCHIEF opened the pull request:
This Pull request removes the `Makefile` and makefile based dependencies, and simplifies the setup and installation steps so that they're the same on any platform.  To begin development, follow the steps now shown in the updated `README.md`

The development of this PR was created during a live-recorded broadcast. You can view this PR in action by visiting [Hacking with Electron & Mattermost](https://www.youtube.com/watch?v=JEbO7L0hjtI)

This PR closes #10

Commit:
initialize npm

Commit:
add the source author as package contributor

Commit:
update README with new usage instructions

Commit:
add start and build scripts

Commit:
Makefile no longer required

Commit:
adding cli executable to installs

Commit:
add npm-debug to .gitignore

Commit:
update path for start and build

Commit:
getting ready for configs

LongLiveCHIEF committed:
initialize vagrant for easy testing server

LongLiveCHIEF committed:
merge testEnv

LongLiveCHIEF committed:
fix for npm-debug

LongLiveCHIEF committed:
adding in auto-launching via npm

you no longer need a makefile or are required to place a config file
in architecture/OS specific folders to run the application in developer
mode. Simply run

    npm install
    npm link
    npm start

in order to launch the app (after adding your config.json). The /usr/local/bin/matterfront -> /usr/local/lib/node_modules/matterfront/bin/cli.js
/usr/local/lib/node_modules/matterfront -> /Users/chief/code/matterfront
command only needs to be run once.

LongLiveCHIEF committed:
update readme with new setup steps

LongLiveCHIEF committed:
removing bin for future PR

LongLiveCHIEF commented:
The first half of the video mentioned in my PR comments is setup of the development Mattermost server via the `Vagrantfile` I've added to the project. For the interesting bits, on how I stand this up without using `make` jump to the [41 minute mark](https://youtu.be/JEbO7L0hjtI?t=2460)

lloeki commented:
Great! I'll review this next.

lloeki commented:
Can you rebase though?

LongLiveCHIEF commented:
sure. I actually have a few things I want to finish before this pull is accepted. So we'll judge based on the completion of the following:
- [x] rebase branch onto existing master
- [x] update README with info on how to modify the `build` script to select platforms of your choice
- [x] finish the Vagrant provisioning script for automatic launch of a local mattermost service for testing
- [x] fix formatting of install instructions in README.md
- [x] Change the `nosrc.html` instructions, no longer need the plat specific stuff

LongLiveCHIEF committed:
merging cross-platform builds

LongLiveCHIEF committed:
file not needed

LongLiveCHIEF committed:
build all plats and archs

LongLiveCHIEF committed:
update README with new build info

LongLiveCHIEF committed:
fix formatting

LongLiveCHIEF committed:
update instructions in

LongLiveCHIEF committed:
update provisioning of test env

LongLiveCHIEF committed:
change to docker provisioner

LongLiveCHIEF committed:
working test mattermost image

LongLiveCHIEF committed:
add local mattermost info to testing docs

LongLiveCHIEF commented:
Ok, this is ready to merge.

Lots of new info in the readme. Also tossed in a self-provisioning local test mattermost instance. Also helps with being able to develop without needing a cloud server and a mattermost install running. This uses the `mattermost/platform:dev` docker image within the test environment, so you'll always have the latest `master` build of the docker image that was pushed to test and develop against.

lloeki commented on README.md:
```diff
@@ -4,15 +4,23 @@ An application for [Mattermost](http://mattermost.org) for OS X, Windows, and Li
 
-- `~/Library/Application Support/Mattermost` on OS X
-- `%APPDATA%/Mattermost` on Windows
-- `~/.config/Mattermost` on Linux
```
Don't remove this yet, as we don't have a UI for configuration.

LongLiveCHIEF replied:
This is no longer needed with the new install and build system. Everything is run from node when in developer mode, and is run via the executable when built. In both scenarios, all that's required is the `/config.json` file.

This was my primary reason for this build... so that instructions config would be the same on any platform for both test/dev as well as build.

See it in action: https://www.youtube.com/watch?v=zY_eu7UIqBE

lloeki commented on src/nosrc.html:
```diff
@@ -1,16 +1,8 @@
-- %APPDATA%/Matterfront
-- ~/Library/Application Support/Matterfront
-- ~/.config/Matterfront
-</code>
```
Don't remove this yet, as we don't have a UI for configuration.

lloeki commented on package.json:
```diff
@@ -0,0 +1,32 @@
+  "name": "matterfront",
+  "version": "0.1.0",
+  "description": "Mattermost frontend app powered by electron",
+  "main": "./src/main.js",
```
This should be `main.js`, see below.

lloeki commented on package.json:
```diff
@@ -0,0 +1,32 @@
+  },
+  "electronVersion":"0.34.1",
+  "scripts": {
+    "start":"electron ./",
```
This should be `src`, as I want to be able to run, build and release an unconfigured app as well as a configured one. This also makes npm start and a packaged app behave the same WRT folder hierarchy (where `src` == app.asar).

H3Chief replied:
Not sure what you mean by this one. Can you hop on the mattermost channel to discuss sometime today?  (invite link below)

LongLiveCHIEF replied:
the `main` property is a hook into node modules system, must be a fully qualified path. It's not a structured run component that's paired with the value passed to `electron` via the `start` script.

the `main` is the file that would be required in if your module was used programmatically.

lloeki replied:
Ok I think I get it now (not used to node idioms), but that's probably my fault for having a `src` dir there. Feels weird to have this asymmetry between `app.asar` and `src`, making the config file lookup blow up in some cases since `app.getAppPath()` doesn't mean the same thing relative to `main.js`.

LongLiveCHIEF replied:
nah, the `/src` dir is pretty common, even in electron apps. You did just fine as far as the node idioms go. :smile: The confusing stuff usually goes away soon, when you start to do more things inside the module's environment itself and get away from the basic system we have now where config for both build and run are coupled together.

I'm working through a heirarchical config system for `build` and `run` type operations that I'll submit a PR for either tonight or tomorrow... which will also make all that stuff _way_ easier to reason about from within the `src` itself.  You'll be able to just make a reference to the app's config, which will be reasoned about based on how it was built _and_ launched. (Going to do some cool stuff with [nconf](https://www.npmjs.com/package/nconf)

lloeki commented on package.json:
```diff
@@ -0,0 +1,32 @@
+  "electronVersion":"0.34.1",
+  "scripts": {
+    "start":"electron ./",
+    "build": "electron-packager ./src matterfront --out=dist --ignore='^/dist$' --prune --asar --platform=all --arch=all --version=$npm_package_electronVersion"
```
Missing some flags: `--icon=resources/mattermost --app-bundle-id=org.matterfront.app --app-version=$npm_package_version --helper-bundle-id=org.matterfront.app.helper --overwrite`

lloeki replied:
Sadly, this outputs `matterfront.app` on OS X, when we'd like to have `Matterfront.app`, but putting `Matterfront` there makes it capitalized for linux. Unless there's a trick in electron-packager to do some magic, I'd rather have it capitalized, as the linux guys will figure this out.

LongLiveCHIEF replied:
I'm planning on writing a more comprehensive build system tomorrow using `electron-packager` 's [programmatic api](https://www.npmjs.com/package/electron-packager#programmatic-api), but for now wanted to get the capability into the build for those waiting for some sort of cross-platform support.

There's enough going into that work that I wanted to make it it's own feature/PR. Felt like this would grow out-of-scope otherwise.

the issues mentioned (case issues, resources/flags) will be fixable from there.

lloeki commented on README.md:
```diff
@@ -27,18 +35,40 @@ In the future, this will be settable from the UI.
+Each directory contains an executable for the platform listed. For more detailed 
+build options, check out how to modify your `build` from within `package.json` by
+using the options from [electron-packager](https://github.com/maxogden/electron-packager#usage) to modify your built artifacts.
 ## Name and affiliation
```
Newline before header

lloeki commented on README.md:
```diff
@@ -27,18 +35,40 @@ In the future, this will be settable from the UI.
+
+### Use your production Mattermost Instance
+
+1. after following the Configuration steps above, run `npm start` from within your matterfront directory 
```
Remove extra `1.`

lloeki commented on README.md:
```diff
@@ -27,18 +35,40 @@ In the future, this will be settable from the UI.
+1. Set up your `config.json` using `"url":"http://192.168.33.33"` 
+2. run `vagrant up` from within your local
+copy of this repo
+2. run `npm start`
```
`3.`

lloeki replied:
Be consistent with caps.

lloeki commented:
This looks nice overall, with just a few details to fix.

LongLiveCHIEF commented:
I'll get on the fixes you found in the morning. Do you want me to work in the build system enhancements with this PR, or do you want to make that a separate PR?

(note: that PR would be the first in which this would be a publish module at npmjs.org, and would have it's own command line utility for launching and parameterizing the launch of Matterfront)

Your call.

Fixes todo:
- [x] consistent caps
- [x] fix bullet numbering
- [x] newline formatting fix
- [x] fix launch dir and associated tips/docs in the `npm start` script
- [x] obtain signoff from @lloeki on no longer requiring platform specific config in `README.md` and `src/nosrc.html`

H3Chief commented:
@lloeki can you hop on the Mattermost core team mattermost channel? [Here's an invite link](https://pre-release.mattermost.com/signup_user_complete/?id=rcgiyftm7jyrxnma1osd8zswby)

LongLiveCHIEF committed:
cleanup README

- fixed formatting issues
- fixed bullet point numbering

LongLiveCHIEF committed:
adding packaging build flags

added flags for , ,
and added an  flag to foce rebuild even if build artifacts
already exist in build directory

LongLiveCHIEF commented:
I added all of the additional build flags, with the exception of `--icon`, which for some reason causes a build failure for the `darwin' platform.  I might have to create a build separation for mac like so:

```
"build:linux": "electron-packager ....",
"build:win": "electron-packager ...",
"build:mac":"electron-packager...",
"build":"npm run build:linux & npm run build:win & npm run build:mac"
```

But at this point, I think I'm just going to go ahead and write out the optional config system as part of this PR

LongLiveCHIEF commented:
Closing this PR before it gets too long. Making some updates to build and will resubmit as new PR.

LongLiveCHIEF committed:
update package.json

- changed package.json props to refer to `lloeki/matterfront`
- rearranged package.json putting basic meta near the top

LongLiveCHIEF committed:
removing package.main

this prop is only needed for modules used programmatically

lloeki commented:
Fine with a new PR, as I feel there are some rough edges that'll be softened up with a couple more things you mentioned.

LongLiveCHIEF committed:
finalize package meta url's
